store/tikv/prewrite.go

```diff
@@ -204,6 +225,29 @@ func (action actionPrewrite) handleSingleBatch(c *twoPhaseCommitter, bo *Backoff
 				key := alreadyExist.GetKey()
 				return c.extractKeyExistsErr(key)
 			}
+			// Handles CommitTsTooLarge error in async commit or 1PC
+			if tooLarge := keyErr.GetCommitTsTooLarge(); tooLarge != nil {
+				logutil.BgLogger().Info("calculated commit TS is too large, fallback to traditional 2PC",
+					zap.Uint64("conn", c.connID),
+					zap.Uint64("txnStartTS", c.startTS),
+					zap.Uint64("allowedMaxCommitTS", c.maxCommitTS),
+					zap.Uint64("calculatedCommitTS", tooLarge.CommitTs))
+				c.setAsyncCommit(false)
+				c.setOnePC(false)
+				newLockTTL := c.refreshLockTTL()
+				// Rewrite the request
+				prewriteReq := req.Req.(*pb.PrewriteRequest)
+				prewriteReq.UseAsyncCommit = false
+				prewriteReq.TryOnePc = false
+				prewriteReq.Secondaries = nil
+				prewriteReq.MaxCommitTs = 0
+				prewriteReq.LockTtl = newLockTTL
+				// if this is not a primary batch, we also need to overwrite the primary lock
+				if !batch.isPrimary {
+					c.reprewritePrimary(bo)
```

Comment:
Seems it's not necessary? Lock resolve can handle all cases now.

Reply:
Sorry I don't understand it. If the primary lock is not changed to non-async-commit, then the primary lock cannot be cleaned by `CheckTxnStatus`.